test/KQStream.ts

```diff
@@ -1,171 +1,155 @@
-            playerKillEvents = [];
-
-            // Act
-            for (let event of events) {
-                connection.sendUTF(event.message);
+            receivedEvents = {
+                playernames: [],
+                playerKill: []
+            };
+
+            const cab = new KQCab();
+            await stream.connect(`ws://localhost:${KQCab.DEFAULT_PORT}`);
+
+            for (let message of messages) {
+                await new Promise<void>((resolve) => {
+                    stream.once(message.type, (arg) => {
+                        (receivedEvents[message.type] as any).push(arg);
+                        resolve();
+                    });
+                    cab.send(message.data);
+                });
             }
-            await sleep(1000);
         });
 
-        it('should process playernames events', () => {
-            expect(playerNamesEvents.length).to.equal(1);
-            for (let expectedEvent of expectedPlayerNamesEvents) {
-                expect(playerNamesEvents).to.deep.include(expectedEvent);
+        it('should process playernames events', async () => {
+            expect(receivedEvents.playernames.length).to.equal(1);
+            for (let expectedEvent of expectedEvents.playernames) {
+                expect(receivedEvents.playernames).to.deep.include(expectedEvent);
             }
         });
 
-        it('should process playerKill events', () => {
-            expect(playerKillEvents.length).to.equal(4);
-            for (let expectedEvent of expectedPlayerKillEvents) {
-                expect(playerKillEvents).to.deep.include(expectedEvent);
+        it('should process playerKill events', async () => {
+            expect(receivedEvents.playerKill.length).to.equal(4);
+            for (let expectedEvent of expectedEvents.playerKill) {
+                expect(receivedEvents.playerKill).to.deep.include(expectedEvent);
             }
         });
+
+        after(() => {
+            stream.removeAllListeners();
+        });
     });
 
     describe('#read', () => {
         before(async () => {
-            // Arrange
-            playerNamesEvents = [];
-            playerKillEvents = [];
+            receivedEvents = {
+                playernames: [],
+                playerKill: []
+            };
+
+            stream.on('playernames', (arg) => {
+                receivedEvents.playernames.push(arg);
+            });
+            stream.on('playerKill', (arg) => {
+                receivedEvents.playerKill.push(arg);
+            });
+
             let data: string = '';
-            for (let event of events) {
-                data += `${event.timestamp},${event.message}\n`;
+            for (let message of messages) {
+                data += `${message.timestamp},${message.data}\n`;
             }
+            stream.read(data);
 
-            // Act
-            kqstream.read(data);
-            await sleep(1000);
+            await sleep(100);
```

Comment:
This is code smell. Find a better way to know that messages have arrived.

Reply:
Also, consider deprecating `KQStream#read`. One can make the argument that KQStream is designed to only receive messages from a Killer Queen cab via websocket connection. If this is the case, why are we letting people read in some alternative format?

Using `KQCab` is the alternative, but it does add an additional layer of complexity by creating a websocket server. However, this is how a connection to a real Killer Queen cabinet would be, and isn't that what we want our code to work with? Isn't using `#read` misleading as it may cause us to overlook issues with `#connect`?

But then again, that's why we unit test both methods...

Give it some consideration.